src/main/java/org/cryptomator/linux/quickaccess/DolphinPlaces.java

```diff
@@ -0,0 +1,161 @@
+	private static final Lock MODIFY_LOCK = new ReentrantReadWriteLock().writeLock();
+	private static final String ENTRY_TEMPLATE = """
+			<bookmark href="%s">
+			 <title>%s</title>
+			 <info>
+			  <metadata owner="http://freedesktop.org">
+			   <bookmark:icon name="drive-harddisk"/>
+			  </metadata>
+			  <metadata owner="https://cryptomator.org">
+			  	<id>%s</id>
+			  </metadata>
+			 </info>
+			</bookmark>""";
+
+
+	private static final Validator xmlValidator;
+
+	static {
+		SchemaFactory factory = SchemaFactory.newInstance(XMLConstants.W3C_XML_SCHEMA_NS_URI);
+		try (var schemaDefinition = DolphinPlaces.class.getResourceAsStream("/xbel-1.0.xsd")) {
+			Source schemaFile = new StreamSource(schemaDefinition);
+			xmlValidator = factory.newSchema(schemaFile).newValidator();
+		} catch (IOException | SAXException e) {
+			throw new IllegalStateException("Failed to load included XBEL schema definition file.", e);
+		}
+	}
+
+
+	@Override
+	public QuickAccessService.QuickAccessEntry add(Path target, String displayName) throws QuickAccessServiceException {
+		String id = UUID.randomUUID().toString();
+		try {
+			MODIFY_LOCK.lock();
+			if (Files.size(PLACES_FILE) > MAX_FILE_SIZE) {
+				throw new IOException("File %s exceeds size of %d bytes".formatted(PLACES_FILE, MAX_FILE_SIZE));
+			}
+			var placesContent = Files.readString(PLACES_FILE);
+			//validate
+			xmlValidator.validate(new StreamSource(new StringReader(placesContent)));
+			// modify
+			int insertIndex = placesContent.lastIndexOf("</xbel>"); //cannot be -1 due to validation
+			try (var writer = Files.newBufferedWriter(TMP_FILE, StandardCharsets.UTF_8, StandardOpenOption.WRITE, StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING)) {
+				writer.write(placesContent, 0, insertIndex);
+				writer.newLine();
+				writer.write(ENTRY_TEMPLATE.formatted(target.toUri(), displayName, id).indent(1));
+				writer.newLine();
+				writer.write(placesContent, insertIndex, placesContent.length() - insertIndex);
+			}
+			// save
+			Files.move(TMP_FILE, PLACES_FILE, StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.ATOMIC_MOVE);
+			return new DolphinPlacesEntry(id);
+		} catch (SAXException | IOException e) {
+			throw new QuickAccessServiceException("Adding entry to KDE places file failed.", e);
+		} finally {
+			MODIFY_LOCK.unlock();
+		}
+	}
+
+	private static class DolphinPlacesEntry implements QuickAccessEntry {
+
+		private final String id;
+		private volatile boolean isRemoved = false;
+
+		DolphinPlacesEntry(String id) {
+			this.id = id;
+		}
+
+		@Override
+		public void remove() throws QuickAccessServiceException {
+			try {
+				MODIFY_LOCK.lock();
+				if (isRemoved) {
+					return;
+				}
+				if (Files.size(PLACES_FILE) > MAX_FILE_SIZE) {
+					throw new IOException("File %s exceeds size of %d bytes".formatted(PLACES_FILE, MAX_FILE_SIZE));
+				}
+				var placesContent = Files.readString(PLACES_FILE);
```

Comment:
Same here: What happens if the file was deleted?

Reply:
An IOException is thrown, which is wrapped in a `QuickAccessServiceException`, as specified by the interface.